### ISSUE TITLE
我感觉这个语言可以再改进一下，把缩进以外的空格全都去掉，这样就完美了。

### ISSUE THREAD

**Issue by vczh** [zhanyong-wan/dongbei#115]
可以参考一下我以前的小玩具：https://github.com/vczh/tinymoe

具体的技术细节我已经实践过了，肯定是没问题的。

🤪

**Comment by zhanyong-wan** [comment]
这个语言里所有字符串之外的空格都是忽略的，包括缩进在内。你发现有什么地方空格是必须的吗？

**Comment by nobodxbodon** [comment]
之前提过 [为识别带有关键词的标识符，可以尝试按语法分词](https://github.com/zhanyong-wan/dongbei/issues/112#issue-1014748011)，不知与 tinymoe 的解析机制是否有相似之处。

**Comment by zhanyong-wan** [comment]
是的。文档里说了： https://github.com/zhanyong-wan/dongbei/blob/master/doc/dongbei-ref.md#%E5%88%86%E8%AF%8D

**Comment by vczh** [comment]
@nobodxbodon 看了那个文章，感觉应该是完全不一样的想法。tinymoe同一句话放在不同的上下文里面，parse出来可能是完全不同的AST，这基本取决于你在上下文里面定义了多少不同的“短语”。举个例子，`print the number`。

如果上下文定义了`print (x)`函数和`the number`变量，那他就是`print (the_number)`。

你甚至可以，定义一个`(x) the number`函数，x是一个functor，那他就变成了 `__the_number (print)`。

你也可以定义一个函数，他就叫`print the number`，那这句话的意思就变成了`print_the_number()`。

**Comment by nobodxbodon** [comment]
@vczh 谢谢介绍。之前个人的实现基本是深度优先搜索找出首个符合语法规则的分词方式，虽有缺陷不过应付比较简单且嵌套较少的语法规则似乎还可以。

tinymoe 这样结合上下文语法元素进行分词感觉比较理想，不过会否有语句A分析依赖于语句B分析但B又依赖A的情况？

**Comment by nobodxbodon** [comment]
@vczh 没理解错的话，声明不会有歧义，语句才会有？

个人感觉，如果是接近自然语言风格的代码，在可读性较好时出现歧义的情况并不太多，歧义处理规则只要能覆盖常用代码，开发者的使用体验就不会太差。在使用中如发现严重歧义的情况，再考虑处理方法不迟。

话说，年前作了点将反馈信息语法和代码语法一致化的设计尝试（[流水在此](https://www.zhihu.com/question/349028712/answer/2220729987)），有空的话烦请批评一下？

**Comment by vczh** [comment]
@nobodxbodon

我还没怎么仔细考虑过你帖子里面的问题，感觉如何表达一个query，需要下很多功夫。特别是编程语言到处都是嵌套的从句，但是中文并不喜欢。

**Comment by nobodxbodon** [comment]
@vczh query是指那个sql领域专用语言原型吗？之前暂用了这种语法（[详见源码](https://github.com/nobodxbodon/rply/blob/5aabd499bcdf7d5f6edfec67cb8955c0c32b459d/tests/test_%E6%8C%89%E8%AF%AD%E6%B3%95%E5%88%86%E8%AF%8D.py#L654)）：
```
获取出生年大于2000的读者的邮箱、昵称，按出生年倒序排列。
```
不过楼上提到的设计是 [另外这个](https://github.com/nobodxbodon/SomeLanguage)：
```
> 摄氏度是180
摄氏度是180

> 华氏度是摄氏度乘9除以5加32
华氏度是摄氏度乘9除以5加32，华氏度是356

> 摄氏度是200
摄氏度是200，华氏度是392
```

特色是反馈信息语法和代码语法保持一致。

**Comment by nobodxbodon** [comment]
> 是的。文档里说了： https://github.com/zhanyong-wan/dongbei/blob/master/doc/dongbei-ref.md#%E5%88%86%E8%AF%8D

想起这番探讨，对[名字包含关键词时需用【】标注](https://github.com/zhanyong-wan/dongbei/blob/master/doc/dongbei-ref.md#%E5%90%8D%E5%AD%97)的限制，[之前的演示](https://github.com/zhanyong-wan/dongbei/issues/112#issue-1014748011)仍有效。

祝2026安康顺遂。
